test/HooglePlus/FilterTestSpec.hs

```diff
@@ -35,6 +38,41 @@ testCases =
   , ("Fail on invalid function 5", ["Data.Maybe"], "Either a b -> Maybe a", "\\x -> fromJust Nothing", False)
   , ("Pass w/ type class 1", [], "(Show a, Show b) => Either a b -> String", "\\x -> show x", True)]
 
+runDuplicateTest :: FilterState -> [String] -> String -> String -> IO (Bool, FilterState)
+runDuplicateTest st modules' funcSig body =
+  runStateT (checkDuplicates (modules ++ modules') funcSig body) st
+
 spec :: Spec
 spec =
-  describe "Filter" $ mapM_ itCase testCases
\ No newline at end of file
+  describe "Filter" $ do
+    mapM_ itNotCrashCase testNotCrashCases
+
+    it "Duplicates - pass on base case" $ do
+      (ret, (FilterState {sampleResults = sample})) <- runDuplicateTest emptyFilterState [] "a -> a" "\\x -> x"
+      ret `shouldBe` True
+      sample `shouldNotBe` Nothing
+
+      let Just (SampleResult inputs outputs) = sample
+      length inputs `shouldBe` defaultNumChecks
+      length outputs `shouldBe` 1
+      
+      let evalResults = head outputs
+      length evalResults `shouldBe` defaultNumChecks
+
+    it "Duplicates - reject identical solution" $ do
+      (ret, st) <- runDuplicateTest emptyFilterState [] "Num a => a -> a" "\\x -> x + 1"
+      (ret', st') <- runDuplicateTest st [] "Num a => a -> a" "\\x -> 1 + x"
+      
+      ret `shouldBe` True
+      ret' `shouldBe` False
+
+      (st == st') `shouldBe` True 
```

Comment:
`shouldBe` doesn't have to take just bools on both sides. You can use `shouldBe st st'`.

Reply:
`shouldBe` seems to use `show` to compare results; based on my testing,`shouldBe st st'` would give an error stating that `FilterState` does not have a instance of `Show`.

Reply:
I don't see any function types that would prevent us from making the `FilterState` type derive `Showable`. Am I overlooking something?

Reply:
Nope. I once thought that we won't like to make `FilterState` derive `Showable` since it seems meaningless. I'll make changes to reflect that.